MAINT/BUG: Fixups to scalar base classes

eric-wieser opened the pull request:
This seems to be left from when `np.int_` subclassed `int`, which it no longer does.

Change the `if` to an `assert` so that we can catch this type of change in future.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

eric-wieser committed:
MAINT: Remove leftover code which handled integer base classes

This seems to be left from when `np.int_` subclassed `int`, which it no longer does.

Change the `if` to an `assert` so that we can catch this type of change in future.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2563,20 +2563,19 @@ object_arrtype_dealloc(PyObject *v)
-        } \
-        /* now do default conversion */ \
-    }
+    assert (type->tp_bases && (PyTuple_GET_SIZE(type->tp_bases) == 2)); \
```
This assertion is failing for subclasses of `np.float64`. The fact that it is suggests that this code isn't doing what it's supposed to be doing anyway though...

seberg replied:
This is patently scary... Is there a reason it is not hardcoding `sup` directly?

eric-wieser replied:
Agreed, very scary - fixed in the follow-up commit.

seberg replied:
Can remove the `num` completely, as far as I can tell, it basically maps to a known instance?

eric-wieser replied:
It does indeed map to a known type, but getting hold of that type through the template mechanism is harder than grabbing it at runtime. Essentially you're choosing between:
```python
class A(B, C):
    def __new__(cls, ...):
        C.__new__(...)
```
and
```python
class A(B, C):
    def __new__(cls, ...):
        A.__bases__[1].__new__(...)
```
.

Both are completely correct, unlike what was there before which was outright stupid:
```python
class A(B, C):
    def __new__(cls, ...):
        cls.__bases__[1].__new__(...)
```

seberg replied:
Yes, the first seemed nicer to me? although not sure if we have the `PyFloatType` name available easily. Could also think about `__bases__[-1]`. in any case, not sure it would make things better or worse, so OK with just leaving it and not doing the last thing is a nice fix.

eric-wieser replied:
Well, `num` is needed because for `unicode_` and `bytes_` it's `__bases__[0]`, but for `float_` it's `__bases__[1]`.

eric-wieser replied:
I suspect I will need to make further cleanups here to address #15385 anyway, in future PRs.

seberg commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2562,25 +2562,20 @@ object_arrtype_dealloc(PyObject *v)
+    if (robj != NULL) goto finish;        \
+    if (PyTuple_GET_SIZE(args) != 1) return NULL; \
+    PyErr_Clear(); \
+    /* now do default conversion */ \
```
```suggestion
    /* now do default conversion */
```

seberg commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -2562,25 +2562,20 @@ object_arrtype_dealloc(PyObject *v)
-#define _WORKz _WORK(0)
-#define _WORK0
+#define _WORK(cls, num)  \
+    assert (cls.tp_bases && (PyTuple_GET_SIZE(cls.tp_bases) == 2)); \
```
```suggestion
    assert(cls.tp_bases && (PyTuple_GET_SIZE(cls.tp_bases) == 2)); \
```

seberg reviewed (approved):
OK, forget about the small stuff if you like)and just put it in yourself once the test pass. This code is really confusing, but the changes look all correct, and like no actual change in behaviour.

eric-wieser committed:
BUG: Do not dispatch to base classes polymorphically

The previously code was equivalent to something like
```python
class SomeScalar(base1, base2):
    def __new__(cls, *args):
        return cls.__bases__[1].__new__(*args)
```
This is nonsense and can easily cause recursion, for instance if `cls` is a superclass of SomeScalar.
The correct way to spell this is to pin the base class call statically.
```python
class SomeScalar(base1, base2):
    def __new__(cls, *args):
        return SomeScalar.__bases__[1].__new__(*args)
```

Fixes #15395

eric-wieser commented:
Rebased locally with the above changes

mattip commented:
Thanks @eric-wieser, @seberg
